In [1]:
import numpy as np 
import pandas as pd

import re

In [2]:
data = pd.read_csv("C:\\Users\\payas\\Downloads\\Data-cleaning-for-beginners-using-pandas.csv")

In [3]:
data

Index   Age      Salary  Rating       Location  Established Easy Apply
0       0  44.0   $44k-$99k     5.4       India,In         1999       TRUE
1       1  66.0   $55k-$66k     3.5    New York,Ny         2002       TRUE
2       2   NaN   $77k-$89k    -1.0    New York,Ny           -1         -1
3       3  64.0   $44k-$99k     4.4       India In         1988         -1
4       4  25.0   $44k-$99k     6.4  Australia Aus         2002         -1
5       5  44.0   $77k-$89k     1.4       India,In         1999       TRUE
6       6  21.0   $44k-$99k     0.0    New York,Ny           -1         -1
7       7  44.0   $44k-$99k    -1.0  Australia Aus           -1         -1
8       8  35.0   $44k-$99k     5.4    New York,Ny           -1         -1
9       9  22.0   $44k-$99k     7.7       India,In           -1       TRUE
10     10  55.0   $10k-$49k     5.4       India,In         2008       TRUE
11     11  44.0   $10k-$49k     6.7       India,In         2009         -1
12     12   NaN   $44k-$99k     0.0       India,In         1999         -1
13     13  25.0   $44k-$99k    -1.0  Australia Aus         2019       TRUE
14     14  66.0   $44k-$99k     4.0  Australia Aus         2020       TRUE
15     15  44.0  $88k-$101k     3.0  Australia Aus         1999         -1
16     16  19.0   $19k-$40k     4.5       India,In         1984         -1
17     17   NaN   $44k-$99k     5.3    New York,Ny         1943       TRUE
18     18  35.0   $44k-$99k     6.7    New York,Ny         1954       TRUE
19     19  32.0   $44k-$99k     3.3    New York,Ny         1955       TRUE
20     20   NaN   $44k-$99k     5.7    New York,Ny         1944       TRUE
21     21  35.0   $44k-$99k     5.0    New York,Ny         1946         -1
22     22  19.0   $55k-$66k     7.8    New York,Ny         1988       TRUE
23     23   NaN   $44k-$99k     2.4    New York,Ny         1999       TRUE
24     24  13.0   $44k-$99k    -1.0    New York,Ny         1987         -1
25     25  55.0   $44k-$99k     0.0  Australia Aus         1980       TRUE
26     26   NaN   $55k-$66k     NaN       India,In         1934       TRUE
27     27  52.0   $44k-$99k     5.4       India,In         1935         -1
28     28   NaN   $39k-$88k     3.4  Australia Aus         1932         -1

In [4]:
data.head()

Index   Age     Salary  Rating       Location  Established Easy Apply
0      0  44.0  $44k-$99k     5.4       India,In         1999       TRUE
1      1  66.0  $55k-$66k     3.5    New York,Ny         2002       TRUE
2      2   NaN  $77k-$89k    -1.0    New York,Ny           -1         -1
3      3  64.0  $44k-$99k     4.4       India In         1988         -1
4      4  25.0  $44k-$99k     6.4  Australia Aus         2002         -1

In [5]:
data.shape

(29, 7)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29 entries, 0 to 28
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Index        29 non-null     int64  
 1   Age          22 non-null     float64
 2   Salary       29 non-null     object 
 3   Rating       28 non-null     float64
 4   Location     29 non-null     object 
 5   Established  29 non-null     int64  
 6   Easy Apply   29 non-null     object 
dtypes: float64(2), int64(2), object(3)
memory usage: 1.7+ KB


In [7]:
data.columns

Index(['Index', 'Age', 'Salary', 'Rating', 'Location', 'Established',
       'Easy Apply'],
      dtype='object')

In [8]:
data['Index'].nunique()

29

In [9]:
data.drop(columns=['Index'],inplace=True)

In [10]:
data.head()

Age     Salary  Rating       Location  Established Easy Apply
0  44.0  $44k-$99k     5.4       India,In         1999       TRUE
1  66.0  $55k-$66k     3.5    New York,Ny         2002       TRUE
2   NaN  $77k-$89k    -1.0    New York,Ny           -1         -1
3  64.0  $44k-$99k     4.4       India In         1988         -1
4  25.0  $44k-$99k     6.4  Australia Aus         2002         -1

In [11]:
data['Age'].unique()

array([44., 66., nan, 64., 25., 21., 35., 22., 55., 19., 32., 13., 52.])

In [12]:
data['Age'].value_counts()

44.0    5
35.0    3
66.0    2
25.0    2
55.0    2
19.0    2
64.0    1
21.0    1
22.0    1
32.0    1
13.0    1
52.0    1
Name: Age, dtype: int64

In [13]:
data['Salary'].unique()

array(['$44k-$99k', '$55k-$66k', '$77k-$89k', '$10k-$49k', '$88k-$101k',
       '$19k-$40k', '$39k-$88k'], dtype=object)

In [14]:
data['Salary'].value_counts()

$44k-$99k     19
$55k-$66k      3
$77k-$89k      2
$10k-$49k      2
$88k-$101k     1
$19k-$40k      1
$39k-$88k      1
Name: Salary, dtype: int64

In [15]:
data['Location'].unique()

array(['India,In', 'New York,Ny', 'India In', 'Australia Aus'],
      dtype=object)

In [16]:
data['Location'].value_counts()

New York,Ny      12
India,In          9
Australia Aus     7
India In          1
Name: Location, dtype: int64

In [17]:
data['Established'].unique()

array([1999, 2002,   -1, 1988, 2008, 2009, 2019, 2020, 1984, 1943, 1954,
       1955, 1944, 1946, 1987, 1980, 1934, 1935, 1932], dtype=int64)

In [18]:
data['Established'].value_counts()

 1999    5
-1       5
 1988    2
 2002    2
 1955    1
 1935    1
 1934    1
 1980    1
 1987    1
 1946    1
 1944    1
 1943    1
 1954    1
 1984    1
 2020    1
 2019    1
 2009    1
 2008    1
 1932    1
Name: Established, dtype: int64

In [19]:
data['Easy Apply'].unique()

array(['TRUE', '-1'], dtype=object)

In [20]:
data['Easy Apply'].value_counts()

TRUE    15
-1      14
Name: Easy Apply, dtype: int64

In [21]:
data.Age

0     44.0
1     66.0
2      NaN
3     64.0
4     25.0
5     44.0
6     21.0
7     44.0
8     35.0
9     22.0
10    55.0
11    44.0
12     NaN
13    25.0
14    66.0
15    44.0
16    19.0
17     NaN
18    35.0
19    32.0
20     NaN
21    35.0
22    19.0
23     NaN
24    13.0
25    55.0
26     NaN
27    52.0
28     NaN
Name: Age, dtype: float64

In [22]:
avg = data['Age'].mean().sum()

In [23]:
data['Age'] = data['Age'].fillna(avg)

In [24]:
data.head()

Age     Salary  Rating       Location  Established Easy Apply
0  44.000000  $44k-$99k     5.4       India,In         1999       TRUE
1  66.000000  $55k-$66k     3.5    New York,Ny         2002       TRUE
2  39.045455  $77k-$89k    -1.0    New York,Ny           -1         -1
3  64.000000  $44k-$99k     4.4       India In         1988         -1
4  25.000000  $44k-$99k     6.4  Australia Aus         2002         -1

In [25]:
data['Age'] = data['Age'].round(decimals=1)

In [26]:
data.Age

0     44.0
1     66.0
2     39.0
3     64.0
4     25.0
5     44.0
6     21.0
7     44.0
8     35.0
9     22.0
10    55.0
11    44.0
12    39.0
13    25.0
14    66.0
15    44.0
16    19.0
17    39.0
18    35.0
19    32.0
20    39.0
21    35.0
22    19.0
23    39.0
24    13.0
25    55.0
26    39.0
27    52.0
28    39.0
Name: Age, dtype: float64

In [27]:
data['Salary'] = data['Salary'].replace({'\$': ''}, regex=True)

In [28]:
data.Salary

0      44k-99k
1      55k-66k
2      77k-89k
3      44k-99k
4      44k-99k
5      77k-89k
6      44k-99k
7      44k-99k
8      44k-99k
9      44k-99k
10     10k-49k
11     10k-49k
12     44k-99k
13     44k-99k
14     44k-99k
15    88k-101k
16     19k-40k
17     44k-99k
18     44k-99k
19     44k-99k
20     44k-99k
21     44k-99k
22     55k-66k
23     44k-99k
24     44k-99k
25     44k-99k
26     55k-66k
27     44k-99k
28     39k-88k
Name: Salary, dtype: object

In [29]:
data['Salary'].replace({'000': ''}, regex=True, inplace=True)

In [30]:
data.Salary

0      44k-99k
1      55k-66k
2      77k-89k
3      44k-99k
4      44k-99k
5      77k-89k
6      44k-99k
7      44k-99k
8      44k-99k
9      44k-99k
10     10k-49k
11     10k-49k
12     44k-99k
13     44k-99k
14     44k-99k
15    88k-101k
16     19k-40k
17     44k-99k
18     44k-99k
19     44k-99k
20     44k-99k
21     44k-99k
22     55k-66k
23     44k-99k
24     44k-99k
25     44k-99k
26     55k-66k
27     44k-99k
28     39k-88k
Name: Salary, dtype: object

In [31]:
data.head()

Age   Salary  Rating       Location  Established Easy Apply
0  44.0  44k-99k     5.4       India,In         1999       TRUE
1  66.0  55k-66k     3.5    New York,Ny         2002       TRUE
2  39.0  77k-89k    -1.0    New York,Ny           -1         -1
3  64.0  44k-99k     4.4       India In         1988         -1
4  25.0  44k-99k     6.4  Australia Aus         2002         -1

In [32]:
data.Rating

0     5.4
1     3.5
2    -1.0
3     4.4
4     6.4
5     1.4
6     0.0
7    -1.0
8     5.4
9     7.7
10    5.4
11    6.7
12    0.0
13   -1.0
14    4.0
15    3.0
16    4.5
17    5.3
18    6.7
19    3.3
20    5.7
21    5.0
22    7.8
23    2.4
24   -1.0
25    0.0
26    NaN
27    5.4
28    3.4
Name: Rating, dtype: float64

In [33]:
Rating_avg= data['Rating'].mean()

In [34]:
data['Rating'].mean()

3.528571428571429

In [35]:
data['Rating']

0     5.4
1     3.5
2    -1.0
3     4.4
4     6.4
5     1.4
6     0.0
7    -1.0
8     5.4
9     7.7
10    5.4
11    6.7
12    0.0
13   -1.0
14    4.0
15    3.0
16    4.5
17    5.3
18    6.7
19    3.3
20    5.7
21    5.0
22    7.8
23    2.4
24   -1.0
25    0.0
26    NaN
27    5.4
28    3.4
Name: Rating, dtype: float64

In [36]:
data['Rating']=data['Rating'].fillna(Rating_avg)

In [37]:
data.Rating

0     5.400000
1     3.500000
2    -1.000000
3     4.400000
4     6.400000
5     1.400000
6     0.000000
7    -1.000000
8     5.400000
9     7.700000
10    5.400000
11    6.700000
12    0.000000
13   -1.000000
14    4.000000
15    3.000000
16    4.500000
17    5.300000
18    6.700000
19    3.300000
20    5.700000
21    5.000000
22    7.800000
23    2.400000
24   -1.000000
25    0.000000
26    3.528571
27    5.400000
28    3.400000
Name: Rating, dtype: float64

In [38]:
data['Rating'].replace({-1.0: 3.5}, inplace=True)


In [39]:
data.Rating

0     5.400000
1     3.500000
2     3.500000
3     4.400000
4     6.400000
5     1.400000
6     0.000000
7     3.500000
8     5.400000
9     7.700000
10    5.400000
11    6.700000
12    0.000000
13    3.500000
14    4.000000
15    3.000000
16    4.500000
17    5.300000
18    6.700000
19    3.300000
20    5.700000
21    5.000000
22    7.800000
23    2.400000
24    3.500000
25    0.000000
26    3.528571
27    5.400000
28    3.400000
Name: Rating, dtype: float64

In [40]:
data.head()

Age   Salary  Rating       Location  Established Easy Apply
0  44.0  44k-99k     5.4       India,In         1999       TRUE
1  66.0  55k-66k     3.5    New York,Ny         2002       TRUE
2  39.0  77k-89k     3.5    New York,Ny           -1         -1
3  64.0  44k-99k     4.4       India In         1988         -1
4  25.0  44k-99k     6.4  Australia Aus         2002         -1

In [41]:
data['Location'] = data['Location'].replace({'New York,Ny':'New York',
                                             'India,In':'India', 
                                             'India In':'India', 
                                             'Australia Aus':'Australia'})

In [42]:
data.Location

0         India
1      New York
2      New York
3         India
4     Australia
5         India
6      New York
7     Australia
8      New York
9         India
10        India
11        India
12        India
13    Australia
14    Australia
15    Australia
16        India
17     New York
18     New York
19     New York
20     New York
21     New York
22     New York
23     New York
24     New York
25    Australia
26        India
27        India
28    Australia
Name: Location, dtype: object

In [43]:
data.head()

Age   Salary  Rating   Location  Established Easy Apply
0  44.0  44k-99k     5.4      India         1999       TRUE
1  66.0  55k-66k     3.5   New York         2002       TRUE
2  39.0  77k-89k     3.5   New York           -1         -1
3  64.0  44k-99k     4.4      India         1988         -1
4  25.0  44k-99k     6.4  Australia         2002         -1

In [44]:
data['Established'].mode()

0      -1
1    1999
Name: Established, dtype: int64

In [45]:
data['Established'].replace({-1:1999},inplace=True)

In [46]:
data.Established

0     1999
1     2002
2     1999
3     1988
4     2002
5     1999
6     1999
7     1999
8     1999
9     1999
10    2008
11    2009
12    1999
13    2019
14    2020
15    1999
16    1984
17    1943
18    1954
19    1955
20    1944
21    1946
22    1988
23    1999
24    1987
25    1980
26    1934
27    1935
28    1932
Name: Established, dtype: int64

In [47]:
data.head()

Age   Salary  Rating   Location  Established Easy Apply
0  44.0  44k-99k     5.4      India         1999       TRUE
1  66.0  55k-66k     3.5   New York         2002       TRUE
2  39.0  77k-89k     3.5   New York         1999         -1
3  64.0  44k-99k     4.4      India         1988         -1
4  25.0  44k-99k     6.4  Australia         2002         -1

In [48]:
data.rename(columns={'Easy Apply':'easy_apply'},inplace=True)

In [49]:
data.head()

Age   Salary  Rating   Location  Established easy_apply
0  44.0  44k-99k     5.4      India         1999       TRUE
1  66.0  55k-66k     3.5   New York         2002       TRUE
2  39.0  77k-89k     3.5   New York         1999         -1
3  64.0  44k-99k     4.4      India         1988         -1
4  25.0  44k-99k     6.4  Australia         2002         -1

In [50]:
data['easy_apply'].replace({'-1':'FALSE'},inplace=True)

In [51]:
data.easy_apply

0      TRUE
1      TRUE
2     FALSE
3     FALSE
4     FALSE
5      TRUE
6     FALSE
7     FALSE
8     FALSE
9      TRUE
10     TRUE
11    FALSE
12    FALSE
13     TRUE
14     TRUE
15    FALSE
16    FALSE
17     TRUE
18     TRUE
19     TRUE
20     TRUE
21    FALSE
22     TRUE
23     TRUE
24    FALSE
25     TRUE
26     TRUE
27    FALSE
28    FALSE
Name: easy_apply, dtype: object

In [52]:
data.head()

Age   Salary  Rating   Location  Established easy_apply
0  44.0  44k-99k     5.4      India         1999       TRUE
1  66.0  55k-66k     3.5   New York         2002       TRUE
2  39.0  77k-89k     3.5   New York         1999      FALSE
3  64.0  44k-99k     4.4      India         1988      FALSE
4  25.0  44k-99k     6.4  Australia         2002      FALSE

In [5]:
continous = ['Age','Salary','Rating','Established']

categorical = ['Location','easy_apply']
             

# Chipotle Dataset

In [136]:
import numpy as np 
import pandas as pd

import re

In [137]:
df = pd.read_csv("C:\\Users\\payas\\chipotle - Copy.tsv",sep='\t')

In [138]:
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

In [139]:
copy_df = df.copy()
copy_df


order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

In [140]:
df.isnull()

order_id  quantity  item_name  choice_description  item_price
0        False     False      False                True       False
1        False     False      False               False       False
2        False     False      False               False       False
3        False     False      False                True       False
4        False     False      False               False       False
...        ...       ...        ...                 ...         ...
4617     False     False      False               False       False
4618     False     False      False               False       False
4619     False     False      False               False       False
4620     False     False      False               False       False
4621     False     False      False               False       False

[4622 rows x 5 columns]

# Data Understanding

	order_id: represent the index of each row.
    
    quantity: no of customers
    
    item_name: Names of the food items/dish
    
    choice_description: choosing
    
    item_price: price of the item
    
    
    
    

In [141]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


In [142]:
df.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [143]:
df['order_id'].unique()

array([   1,    2,    3, ..., 1832, 1833, 1834], dtype=int64)

In [144]:
df['order_id'].value_counts()

926     23
1483    14
205     12
759     11
1786    11
        ..
768      1
341      1
1048     1
94       1
1199     1
Name: order_id, Length: 1834, dtype: int64

In [145]:
df['quantity'].unique()

array([ 1,  2,  3,  4,  5, 15,  7,  8, 10], dtype=int64)

In [146]:
df['quantity'].value_counts()

1     4355
2      224
3       28
4       10
5        1
15       1
7        1
8        1
10       1
Name: quantity, dtype: int64

In [147]:
df['item_name'].unique()

array(['Chips and Fresh Tomato Salsa', 'Izze', 'Nantucket Nectar',
       'Chips and Tomatillo-Green Chili Salsa', 'Chicken Bowl',
       'Side of Chips', 'Steak Burrito', 'Steak Soft Tacos',
       'Chips and Guacamole', 'Chicken Crispy Tacos',
       'Chicken Soft Tacos', 'Chicken Burrito', 'Canned Soda',
       'Barbacoa Burrito', 'Carnitas Burrito', 'Carnitas Bowl',
       'Bottled Water', 'Chips and Tomatillo Green Chili Salsa',
       'Barbacoa Bowl', 'Chips', 'Chicken Salad Bowl', 'Steak Bowl',
       'Barbacoa Soft Tacos', 'Veggie Burrito', 'Veggie Bowl',
       'Steak Crispy Tacos', 'Chips and Tomatillo Red Chili Salsa',
       'Barbacoa Crispy Tacos', 'Veggie Salad Bowl',
       'Chips and Roasted Chili-Corn Salsa',
       'Chips and Roasted Chili Corn Salsa', 'Carnitas Soft Tacos',
       'Chicken Salad', 'Canned Soft Drink', 'Steak Salad Bowl',
       '6 Pack Soft Drink', 'Chips and Tomatillo-Red Chili Salsa', 'Bowl',
       'Burrito', 'Crispy Tacos', 'Carnitas Crispy Tacos

In [148]:
df['item_name'].value_counts()

Chicken Bowl                             726
Chicken Burrito                          553
Chips and Guacamole                      479
Steak Burrito                            368
Canned Soft Drink                        301
Steak Bowl                               211
Chips                                    211
Bottled Water                            162
Chicken Soft Tacos                       115
Chips and Fresh Tomato Salsa             110
Chicken Salad Bowl                       110
Canned Soda                              104
Side of Chips                            101
Veggie Burrito                            95
Barbacoa Burrito                          91
Veggie Bowl                               85
Carnitas Bowl                             68
Barbacoa Bowl                             66
Carnitas Burrito                          59
Steak Soft Tacos                          55
6 Pack Soft Drink                         54
Chips and Tomatillo Red Chili Salsa       48
Chicken Cr

In [149]:
df['choice_description'].unique()

array([nan, '[Clementine]', '[Apple]', ...,
       '[Roasted Chili Corn Salsa, [Pinto Beans, Sour Cream, Cheese, Lettuce, Guacamole]]',
       '[Tomatillo Green Chili Salsa, [Rice, Black Beans]]',
       '[Tomatillo Green Chili Salsa, [Rice, Fajita Vegetables, Black Beans, Guacamole]]'],
      dtype=object)

In [150]:
df['choice_description'].value_counts()

[Diet Coke]                                                                                                                                      134
[Coke]                                                                                                                                           123
[Sprite]                                                                                                                                          77
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]                                                                            42
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Guacamole, Lettuce]]                                                                 40
                                                                                                                                                ... 
[Fresh Tomato Salsa (Mild), [Pinto Beans, Black Beans, Rice, Cheese, Sour Cream, Lettuce]]                

In [151]:
df['item_price'].unique()

array(['$2.39 ', '$3.39 ', '$16.98 ', '$10.98 ', '$1.69 ', '$11.75 ',
       '$9.25 ', '$4.45 ', '$8.75 ', '$11.25 ', '$8.49 ', '$2.18 ',
       '$8.99 ', '$1.09 ', '$2.95 ', '$2.15 ', '$3.99 ', '$22.50 ',
       '$11.48 ', '$17.98 ', '$17.50 ', '$4.30 ', '$5.90 ', '$1.25 ',
       '$23.78 ', '$6.49 ', '$11.08 ', '$1.50 ', '$22.16 ', '$32.94 ',
       '$22.20 ', '$10.58 ', '$2.50 ', '$23.50 ', '$7.40 ', '$18.50 ',
       '$3.00 ', '$6.78 ', '$11.89 ', '$9.39 ', '$4.00 ', '$3.75 ',
       '$8.69 ', '$2.29 ', '$8.90 ', '$3.27 ', '$3.89 ', '$8.19 ',
       '$35.00 ', '$27.75 ', '$11.80 ', '$6.00 ', '$26.25 ', '$21.96 ',
       '$4.36 ', '$7.50 ', '$4.78 ', '$13.35 ', '$6.45 ', '$5.07 ',
       '$22.96 ', '$7.17 ', '$7.98 ', '$4.50 ', '$26.07 ', '$12.98 ',
       '$35.25 ', '$44.25 ', '$10.50 ', '$33.75 ', '$16.38 ', '$13.52 ',
       '$5.00 ', '$15.00 ', '$8.50 ', '$17.80 ', '$1.99 ', '$11.49 '],
      dtype=object)

In [152]:
df['item_price'].value_counts()

$8.75      730
$11.25     521
$9.25      398
$4.45      349
$8.49      311
          ... 
$13.35       1
$7.50        1
$4.36        1
$11.80       1
$11.49       1
Name: item_price, Length: 78, dtype: int64

In [153]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [154]:
df['order_id']

0          1
1          1
2          1
3          1
4          2
        ... 
4617    1833
4618    1833
4619    1834
4620    1834
4621    1834
Name: order_id, Length: 4622, dtype: int64

In [155]:
df['quantity'].isna().sum()

0

In [156]:
df['quantity'].isna().sum()

0

In [157]:
df['choice_description'].isna()

0        True
1       False
2       False
3        True
4       False
        ...  
4617    False
4618    False
4619    False
4620    False
4621    False
Name: choice_description, Length: 4622, dtype: bool

In [158]:
df['choice_description'].isna().sum()

1246

In [159]:
df['choice_description'].value_counts()

[Diet Coke]                                                                                                                                      134
[Coke]                                                                                                                                           123
[Sprite]                                                                                                                                          77
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]                                                                            42
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Guacamole, Lettuce]]                                                                 40
                                                                                                                                                ... 
[Fresh Tomato Salsa (Mild), [Pinto Beans, Black Beans, Rice, Cheese, Sour Cream, Lettuce]]                

In [160]:
choice_mode=df['choice_description'].mode().sum()

In [161]:
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

In [162]:
df['choice_description'].fillna((choice_mode),inplace = True)

In [163]:
df.choice_description

0                                             [Diet Coke]
1                                            [Clementine]
2                                                 [Apple]
3                                             [Diet Coke]
4       [Tomatillo-Red Chili Salsa (Hot), [Black Beans...
                              ...                        
4617    [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...
4618    [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...
4619    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
4620    [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...
4621    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
Name: choice_description, Length: 4622, dtype: object

In [164]:
df.choice_description

0                                             [Diet Coke]
1                                            [Clementine]
2                                                 [Apple]
3                                             [Diet Coke]
4       [Tomatillo-Red Chili Salsa (Hot), [Black Beans...
                              ...                        
4617    [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...
4618    [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...
4619    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
4620    [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...
4621    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
Name: choice_description, Length: 4622, dtype: object

In [165]:
df['choice_description'].isnull().sum()

0

In [166]:
df.head(20)

order_id  quantity                              item_name  \
0          1         1           Chips and Fresh Tomato Salsa   
1          1         1                                   Izze   
2          1         1                       Nantucket Nectar   
3          1         1  Chips and Tomatillo-Green Chili Salsa   
4          2         2                           Chicken Bowl   
5          3         1                           Chicken Bowl   
6          3         1                          Side of Chips   
7          4         1                          Steak Burrito   
8          4         1                       Steak Soft Tacos   
9          5         1                          Steak Burrito   
10         5         1                    Chips and Guacamole   
11         6         1                   Chicken Crispy Tacos   
12         6         1                     Chicken Soft Tacos   
13         7         1                           Chicken Bowl   
14         7         1                    Chips and Guacamole   
15         8         1  Chips and Tomatillo-Green Chili Salsa   
16         8         1                        Chicken Burrito   
17         9         1                        Chicken Burrito   
18         9         2                            Canned Soda   
19        10         1                           Chicken Bowl   

                                   choice_description item_price  
0                                         [Diet Coke]     $2.39   
1                                        [Clementine]     $3.39   
2                                             [Apple]     $3.39   
3                                         [Diet Coke]     $2.39   
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                         [Diet Coke]     $1.69   
7   [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8   [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9   [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25   
10                                        [Diet Coke]     $4.45   
11  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75   
12  [Roasted Chili Corn Salsa, [Rice, Black Beans,...     $8.75   
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
14                                        [Diet Coke]     $4.45   
15                                        [Diet Coke]     $2.39   
16  [Tomatillo-Green Chili Salsa (Medium), [Pinto ...     $8.49   
17  [Fresh Tomato Salsa (Mild), [Black Beans, Rice...     $8.49   
18                                           [Sprite]     $2.18   
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...     $8.75

In [167]:
df["item_price"].isna().sum()

0

In [168]:
df['item_name'].isna().sum()

0

In [169]:
df['quantity'].isna().sum()

0

In [170]:
df['item_name'].duplicated().sum()

4572

In [171]:
print(df.duplicated().sum())

59


In [172]:
df.drop_duplicates(inplace = True) 

In [173]:
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                           [Diet Coke]     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                           [Diet Coke]     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4563 rows x 5 columns]

In [174]:
print(df.duplicated().sum())

0


In [175]:
df['choice_description'] = df['choice_description'].str.replace('[', '', regex=True)
df['choice_description'] = df['choice_description'].str.replace(']', '', regex=True)
df['choice_description'] = df['choice_description'].str.replace('(', '', regex=True)
df['choice_description'] = df.choice_description.str.replace("  ", " ")
df['choice_description'] = df['choice_description'].str.replace(')', '', regex=True)

In [176]:
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                             Diet Coke     $2.39   
1                                            Clementine     $3.39   
2                                                 Apple     $3.39   
3                                             Diet Coke     $2.39   
4     Tomatillo-Red Chili Salsa Hot, Black Beans, Ri...    $16.98   
...                                                 ...        ...  
4617  Fresh Tomato Salsa, Rice, Black Beans, Sour Cr...    $11.75   
4618  Fresh Tomato Salsa, Rice, Sour Cream, Cheese, ...    $11.75   
4619  Fresh Tomato Salsa, Fajita Vegetables, Pinto B...    $11.25   
4620     Fresh Tomato Salsa, Fajita Vegetables, Lettuce     $8.75   
4621  Fresh Tomato Salsa, Fajita Vegetables, Pinto B...     $8.75   

[4563 rows x 5 columns]

In [177]:
df['item_price'] = df['item_price'].replace({'\$': ''}, regex=True)

In [178]:
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                             Diet Coke      2.39   
1                                            Clementine      3.39   
2                                                 Apple      3.39   
3                                             Diet Coke      2.39   
4     Tomatillo-Red Chili Salsa Hot, Black Beans, Ri...     16.98   
...                                                 ...        ...  
4617  Fresh Tomato Salsa, Rice, Black Beans, Sour Cr...     11.75   
4618  Fresh Tomato Salsa, Rice, Sour Cream, Cheese, ...     11.75   
4619  Fresh Tomato Salsa, Fajita Vegetables, Pinto B...     11.25   
4620     Fresh Tomato Salsa, Fajita Vegetables, Lettuce      8.75   
4621  Fresh Tomato Salsa, Fajita Vegetables, Pinto B...      8.75   

[4563 rows x 5 columns]

In [201]:
df.describe()

order_id     quantity
count  4563.000000  4563.000000
mean    927.045803     1.076704
std     528.936100     0.412739
min       1.000000     1.000000
25%     476.500000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

In [204]:
df.isnull().sum()

order_id              0
quantity              0
item_name             0
choice_description    0
item_price            0
dtype: int64

In [179]:
item_quantities = df.groupby('item_name')['quantity'].sum()

In [180]:
item_quantities

item_name
6 Pack Soft Drink                         55
Barbacoa Bowl                             65
Barbacoa Burrito                          90
Barbacoa Crispy Tacos                     12
Barbacoa Salad Bowl                        9
Barbacoa Soft Tacos                       25
Bottled Water                            204
Bowl                                       4
Burrito                                    6
Canned Soda                              124
Canned Soft Drink                        340
Carnitas Bowl                             71
Carnitas Burrito                          60
Carnitas Crispy Tacos                      8
Carnitas Salad                             1
Carnitas Salad Bowl                        6
Carnitas Soft Tacos                       40
Chicken Bowl                             752
Chicken Burrito                          584
Chicken Crispy Tacos                      50
Chicken Salad                              9
Chicken Salad Bowl                       123


In [196]:
top_items = item_quantities.sort_values(ascending=False).head()

In [197]:
top_items


item_name
Chicken Bowl           752
Chicken Burrito        584
Chips and Guacamole    501
Steak Burrito          383
Canned Soft Drink      340
Name: quantity, dtype: int64

Q1.Quantity of each item? (Top 5 itmes available in store)

#item_name

1.Chicken Bowl                             752

2.Chicken Burrito                          584

3.Chips and Guacamole                      501

4.Steak Burrito                            383

5.Canned Soft Drink                        340

In [183]:
choice_quantities = df.groupby('choice_description')['quantity'].sum()


In [184]:
choice_quantities 

choice_description
Adobo-Marinated and Grilled Chicken                                                                              1
Adobo-Marinated and Grilled Chicken, Pinto Beans, Sour Cream, Salsa, Cheese, Cilantro-Lime Rice, Guacamole       1
Adobo-Marinated and Grilled Chicken, Sour Cream, Cheese, Cilantro-Lime Rice                                      1
Adobo-Marinated and Grilled Steak                                                                                1
Adobo-Marinated and Grilled Steak, Sour Cream, Salsa, Cheese, Cilantro-Lime Rice, Guacamole                      1
                                                                                                                ..
Tomatillo-Red Chili Salsa Hot, Tomatillo-Green Chili Salsa Medium, Rice, Pinto Beans, Fajita Veggies, Lettuce    1
White Rice                                                                                                       1
White Rice, Adobo-Marinated and Grilled Chicken, Vegetarian B

In [200]:
most_ordered_item = choice_quantities.idxmax()


In [186]:
most_ordered_item

'Diet Coke'

Q2.which is most ordered item in choice_description?

1.Diet Coke

In [187]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                          Diet Coke      2.39   
1                                         Clementine      3.39   
2                                              Apple      3.39   
3                                          Diet Coke      2.39   
4  Tomatillo-Red Chili Salsa Hot, Black Beans, Ri...     16.98

In [188]:
total_order_count = df['order_id'].nunique()

In [189]:
total_order_count

1834

In [190]:
most_ordered_quantity = choice_quantities.max()
most_ordered_quantity

1522

Q3.What is total order count?

 The total_order_count is = 1834

Q4.What is revenue?



In [206]:
total_revenue= (df.item_price * df.quantity).sum()

In [207]:
total_revenue

'2.39 3.39 3.39 2.39 16.98 16.98 10.98 1.69 11.75 9.25 9.25 4.45 8.75 8.75 11.25 4.45 2.39 8.49 8.49 2.18 2.18 8.75 4.45 8.99 3.39 10.98 3.39 2.39 8.49 8.99 1.09 8.49 2.39 8.99 1.69 8.99 1.09 8.75 8.75 4.45 2.95 11.75 2.15 4.45 11.25 11.75 8.75 10.98 8.99 3.39 8.99 3.99 8.99 2.18 2.18 10.98 1.09 8.99 2.39 9.25 11.25 11.75 2.15 4.45 9.25 11.25 8.75 8.99 8.99 3.39 8.99 10.98 8.99 1.69 8.99 3.99 8.75 4.45 8.75 8.75 2.15 8.75 11.25 2.15 9.25 8.75 8.75 9.25 8.49 8.99 1.09 9.25 2.95 11.75 11.75 9.25 11.75 4.45 9.25 4.45 11.75 8.75 8.75 4.45 8.99 8.99 3.99 8.49 3.39 8.99 1.09 9.25 4.45 8.75 2.95 4.45 2.39 8.49 8.99 8.49 1.09 8.99 3.99 8.75 9.25 4.45 11.25 4.45 8.99 1.09 9.25 2.95 4.45 11.75 4.45 8.49 2.39 10.98 22.50 22.50 11.75 4.45 11.25 4.45 11.25 4.45 11.25 11.25 11.75 9.25 4.45 11.48 17.98 17.98 1.69 17.50 17.50 4.45 8.49 2.39 17.50 17.50 4.45 4.45 11.25 11.75 10.98 8.49 10.98 2.18 2.18 11.48 8.49 2.39 4.45 11.25 11.75 8.75 8.49 2.18 2.18 8.49 3.39 8.49 8.99 10.98 11.48 8.49 1.09 1.09 9.